core/store/src/flat_state.rs

```diff
@@ -352,49 +357,130 @@ pub struct FlatStorageState(Arc<RwLock<FlatStorageStateInner>>);
+    blocks: HashMap<CryptoHash, BlockInfo>,
     /// State deltas for all blocks supported by this flat storage.
     /// All these deltas here are stored on disk too.
     #[allow(unused)]
     deltas: HashMap<CryptoHash, FlatStateDelta>,
 }
 
+#[cfg(feature = "protocol_feature_flat_state")]
+pub mod store_helper {
+    use crate::flat_state::KeyForFlatStateDelta;
+    use crate::{FlatStateDelta, Store, StoreUpdate};
+    use borsh::BorshSerialize;
+    use near_primitives::hash::CryptoHash;
+    use near_primitives::types::ShardId;
+
+    pub(crate) fn get_delta(
+        store: &Store,
+        shard_id: ShardId,
+        block_hash: CryptoHash,
+    ) -> Result<Option<FlatStateDelta>, crate::StorageError> {
+        let key = KeyForFlatStateDelta { shard_id, block_hash };
+        store
+            .get_ser(crate::DBCol::FlatStateDeltas, &key.try_to_vec().unwrap())
+            .map_err(|_| crate::StorageError::StorageInternalError)
+    }
+
+    pub fn set_delta(
+        store_update: &mut StoreUpdate,
+        shard_id: ShardId,
+        block_hash: CryptoHash,
+        delta: &FlatStateDelta,
+    ) -> Result<(), crate::StorageError> {
+        let key = KeyForFlatStateDelta { shard_id, block_hash };
+        store_update
+            .set_ser(crate::DBCol::FlatStateDeltas, &key.try_to_vec().unwrap(), delta)
+            .map_err(|_| crate::StorageError::StorageInternalError)
+    }
+
+    pub(crate) fn get_flat_head(store: &Store, shard_id: ShardId) -> CryptoHash {
+        store
+            .get_ser(crate::DBCol::FlatStateMisc, &shard_id.try_to_vec().unwrap())
+            .expect("Error reading flat head from storage")
+            .expect(&format!("Cannot read flat head for shard {} from storage", shard_id))
+    }
+
+    pub fn set_flat_head(store_update: &mut StoreUpdate, shard_id: ShardId, val: &CryptoHash) {
+        store_update
+            .set_ser(crate::DBCol::FlatStateMisc, &shard_id.try_to_vec().unwrap(), val)
+            .expect("Error writing flat head from storage")
+    }
+}
+
 impl FlatStorageState {
-    #[allow(unused)]
+    /// Create a new FlatStorageState for `shard_id`.
+    /// The root block of the FlatStorageState is initialized to be what is stored on storage.
+    /// We also load all blocks with height between the root block to the current chain head,
+    /// including those on forks into the returned FlatStorageState.
+    #[cfg(feature = "protocol_feature_flat_state")]
     pub fn new(
         store: Store,
-        head: CryptoHash,
-        tail: CryptoHash,
-        block_headers: HashMap<CryptoHash, BlockHeader>,
+        chain_head_height: BlockHeight,
+        // Unfortunately we don't have access to ChainStore inside this file because of package
+        // dependencies, so we pass these functions in to access chain info
+        hash_to_header: &dyn Fn(&CryptoHash) -> BlockHeader,
+        height_to_hashes: &dyn Fn(BlockHeight) -> HashSet<CryptoHash>,
```

Comment:
I like that trick! :)
Maybe we can revisit this later, when all the pieces for flat storage state are in place. It could potentially make sense to elevate some shared struct to a shared dependency crate. Or even create a new trait in `core/store` and pass down something like `&dyn ChainStoreView` here. But I wouldn't bother with that for now, passing down of two function pointers is so much simpler.

Reply:
Good idea! I will do that in the next PR since the interface may change.